### PR TITLE
add keepalive roundtripper to vimClient

### PR DESCRIPTION
**Pull request by sadysnaat**
**What this PR does / why we need it**:
This PR adds session.Keepalive to vimclient 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # 

**Special notes for your reviewer**:
Ref PR https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1072 

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/manager/constants.go [hunk 1]
```diff
@@ -29,6 +29,11 @@ const (
 	// manager option.
 	DefaultSyncPeriod = time.Minute * 10
 
+	// DefaultEnableKeepAlive is false by default
+	DefaultEnableKeepAlive = true
+
+	DefaultKeepAliveDuration = time.Duration(5)
```
Comment (jzhoucliqr) [inline pkg/manager/constants.go:35]: shall we be more specific, is this 5 min / 5 sec?

Comment (sadysnaat) [reply pkg/manager/constants.go:36]: Ack. DefaultEnableKeepalive changed to false.

File: pkg/session/session.go [hunk 2]
```diff
@@ -95,28 +127,53 @@ func GetOrCreate(
 	return &session, nil
 }
 
-func newClient(ctx context.Context, url *url.URL, thumprint string) (*govmomi.Client, error) {
+func newClient(ctx GetOrCreateContext, sessionKey string, url *url.URL, thumprint string, feature Feature) (*govmomi.Client, error) {
```
Comment (srm09) [inline pkg/session/session.go:130]: ~~If we are not using the `logger` here, we could just pass the context instead.~~

File: pkg/session/session.go [hunk 3]
```diff
@@ -95,28 +127,53 @@ func GetOrCreate(
...
+	}
+
+	if err := c.Login(ctx.context, url.User); err != nil {
 		return nil, err
 	}
 
 	return c, nil
 }
 
+func ClearCache(sessionKey string) {
```
Comment (srm09) [inline pkg/session/session.go:171]:
```suggestion
func clearCache(sessionKey string) {
```

File: pkg/session/session.go [hunk 4]
```diff
@@ -95,28 +127,53 @@ func GetOrCreate(
...
+			// we tried implementing
+			// c.Login here but the client once logged out
+			// keeps errong in invalid username or password
+			// we tried with cached username and password in session still the error persisted
+			// hence we just clear the cache and expect the client to
+			// be recreated in next GetOrCreate call
+			_, err := methods.GetCurrentTime(ctx.context, tripper)
+			if err != nil {
+				ctx.logger.Error(err, "failed to keep alive govmomi client")
+				ClearCache(sessionKey)
```
Comment (srm09) [inline pkg/session/session.go:158]:
```suggestion
				clearCache(sessionKey)
```

**Review by srm09 — CHANGES_REQUESTED** [review]
Took an initial pass at it. Let me know if you have any questions.

File: pkg/session/session.go [hunk 5]
```diff
@@ -42,18 +45,47 @@ type Session struct {
...
+
+type Feature struct {
+	EnableKeepAlive   bool
+	KeepAliveDuration time.Duration
+}
+
+func DefaultFeature() Feature {
+	return Feature{
+		EnableKeepAlive:   false,
+		KeepAliveDuration: 0,
```
[the remark this answers is not in the document]
Comment (sadysnaat) [reply pkg/session/session.go]: Ack

File: pkg/session/session.go [hunk 6]
```diff
@@ -42,18 +45,47 @@ type Session struct {
 	datacenter *object.Datacenter
 }
 
+type GetOrCreateContext struct {
+	context context.Context
+	logger  logr.Logger
+}
```
[the remark this answers is not in the document]
Comment (sadysnaat) [reply pkg/session/session.go]: This struct is just a wrapper I have created. could you please point me to PR

File: pkg/session/session.go [hunk 7]
```diff
@@ -42,18 +45,47 @@ type Session struct {
...
+	}
+}
+
 // GetOrCreate gets a cached session or creates a new one if one does not
 // already exist.
 func GetOrCreate(
-	ctx context.Context,
-	server, datacenter, username, password string, thumbprint string) (*Session, error) {
+	ctx GetOrCreateContext,
+	server, datacenter, username, password string, thumbprint string, feature Feature) (*Session, error) {
```
[the remark this answers is not in the document]
Comment (sadysnaat) [reply pkg/session/session.go]: Ack. I have used url.UserInfo for username, wdyt? Should I use a struct instead?

File: pkg/session/session.go (hunk 3 above)
Comment (sadysnaat) [reply pkg/session/session.go]: Ack

File: pkg/session/session.go (hunk 4 above)
Comment (sadysnaat) [reply pkg/session/session.go]: Ack

File: pkg/session/session.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (srm09) [reply pkg/session/session.go]: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1138 has been merged to the `main` branch now.

**Review by srm09 — COMMENTED** [review]
minor nits, added the PR number for the controller-runtime change. Lemme know if it makes sense.

File: pkg/session/session.go (hunk 3 above)
Comment (srm09) [reply pkg/session/session.go]: Disregard this one, we do not wanna expose the clearCache method outside the package.

File: pkg/session/session.go (hunk 4 above)
Comment (srm09) [reply pkg/session/session.go]: Same as above. disregard this one.

File: pkg/session/session.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (srm09) [reply pkg/session/session.go]: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1171/files#diff-59fbb0bb5dfeeb4f123f1e9e0a3d12c38a60b839e113b29868cb905755639e13R51
This change fetches the logger from the context.

File: pkg/services/govmomi/vcenter/clone_test.go [hunk 8]
```diff
@@ -141,10 +142,18 @@ func initSimulator(t *testing.T) (*simulator.Model, *session.Session, *simulator
...
 	authSession, err := session.GetOrCreate(
-		ctx.TODO(),
-		server.URL.Host, "",
-		server.URL.User.Username(), pass, "")
+		session.NewSessionContext(ctx.TODO(), klogr.New()),
+		session.NewParams().
+			WithServer(server.URL.Host).
+			WithDatacenter("").
+			WithUserInfo(server.URL.User.Username(), pass).
+			WithThumbprint(""))
```
[the remark this answers is not in the document]
Comment (sadysnaat) [reply pkg/services/govmomi/vcenter/clone_test.go]: Ack

File: pkg/services/govmomi/vcenter/clone_test.go [hunk 9]
```diff
@@ -141,10 +142,18 @@ func initSimulator(t *testing.T) (*simulator.Model, *session.Session, *simulator
...
+	//	server.URL.User.Username(), pass, "",
+	//	session.DefaultFeature())
 	authSession, err := session.GetOrCreate(
-		ctx.TODO(),
-		server.URL.Host, "",
-		server.URL.User.Username(), pass, "")
+		session.NewSessionContext(ctx.TODO(), klogr.New()),
+		session.NewParams().
+			WithServer(server.URL.Host).
+			WithDatacenter("").
```
[the remark this answers is not in the document]
Comment (sadysnaat) [reply pkg/services/govmomi/vcenter/clone_test.go]: Ack

**Comment by sadysnaat** [comment]
@yastij @srm09 Not sure if my latest changes are causing e2e test failure 

```
Failure [4.479 seconds]
[BeforeSuite] BeforeSuite 
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-vsphere/test/e2e/e2e_suite_test.go:104
  Unexpected error:
      <soap.soapFaultError>: {
          fault: {
              XMLName: {
                  Space: "http://schemas.xmlsoap.org/soap/envelope/",
                  Local: "Fault",
              },
              Code: "ServerFaultCode",
              String: "Cannot complete login due to an incorrect user name or password.",
              Detail: {
                  Fault: {
                      VimFault: {
                          MethodFault: {FaultCause: nil, FaultMessage: nil},
                      },
                  },
              },
          },
      }
      ServerFaultCode: Cannot complete login due to an incorrect user name or password.
  occurred
  /home/prow/go/src/sigs.k8s.io/cluster-api-provider-vsphere/test/e2e/govmomi_test.go:58
------------------------------
STEP: Tearing down the management cluster
JUnit report was created: /logs/artifacts/junit.e2e_suite.1.xml
Ran 2 of 0 Specs in 4.479 seconds
FAIL! -- 0 Passed | 2 Failed | 0 Pending | 0 Skipped
--- FAIL: TestE2E (4.48s)
FAIL 
```

File: pkg/context/controller_manager_context.go [hunk 10]
```diff
@@ -19,14 +19,14 @@ package context
 import (
 	"context"
 	"sync"
+	"time"
 
 	"github.com/go-logr/logr"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/record"
```
Comment (srm09) [inline pkg/context/controller_manager_context.go:27]: nit: Please revert this change. The way the imports are organized are:
```
golang native packages
<empty line>
third party packages
<empty line>
imports from the same project
```

**Comment by srm09** [comment]
@sadysnaat Ran this test locally and it did work. Am unsure as to why this wouldn't work.

/test pull-cluster-api-provider-vsphere-e2e

**Comment by srm09** [comment]
@yastij Do you mind taking a look at this one?

**Review by yastij — COMMENTED** [review]
/approve

lgtm pending rebase
